thrifty-runtime/src/test/java/com/microsoft/thrifty/protocol/JsonProtocolTest.java

```diff
@@ -0,0 +1,195 @@
+    }
+
+    @Test
+    public void listWithOneElement() throws Exception {
+        protocol.writeListBegin(TType.STRING, 1);
+        protocol.writeString("foo");
+        protocol.writeListEnd();
+
+        assertThat(buffer.readUtf8()).isEqualTo("[\"str\",1,\"foo\"]");
+    }
+
+    @Test
+    public void listWithTwoElements() throws Exception {
+        protocol.writeListBegin(TType.STRING, 2);
+        protocol.writeString("foo");
+        protocol.writeString("bar");
+        protocol.writeListEnd();
+
+        assertThat(buffer.readUtf8()).isEqualTo("[\"str\",2,\"foo\",\"bar\"]");
+    }
+
+    @Test
+    public void emptyMap() throws Exception {
+        protocol.writeMapBegin(TType.STRING, TType.I32, 0);
+        protocol.writeMapEnd();
+
+        assertThat(buffer.readUtf8()).isEqualTo("[\"str\",\"i32\",0,{}]");
+    }
+
+    @Test
+    public void mapWithSingleElement() throws Exception {
+        protocol.writeMapBegin(TType.STRING, TType.I32, 1);
+        protocol.writeString("key1");
+        protocol.writeI32(1);
+        protocol.writeMapEnd();
+
+        assertThat(buffer.readUtf8()).isEqualTo("[\"str\",\"i32\",1,{\"key1\":1}]");
+    }
+
+    @Test
+    public void mapWithTwoElements() throws Exception {
+        protocol.writeMapBegin(TType.STRING, TType.I32, 2);
+        protocol.writeString("key1");
+        protocol.writeI32(1);
+        protocol.writeString("key2");
+        protocol.writeI32(2);
+        protocol.writeMapEnd();
+
+        assertThat(buffer.readUtf8()).isEqualTo("[\"str\",\"i32\",2,{\"key1\":1,\"key2\":2}]");
+    }
+
+    @Test
+    public void listOfMaps() throws Exception {
+        protocol.writeListBegin(TType.MAP, 2);
+
+        protocol.writeMapBegin(TType.STRING, TType.I32, 1);
+        protocol.writeString("1");
+        protocol.writeI32(10);
+        protocol.writeMapEnd();
+
+        protocol.writeMapBegin(TType.STRING, TType.I32, 1);
+        protocol.writeString("2");
+        protocol.writeI32(20);
+        protocol.writeMapEnd();
+
+        protocol.writeListEnd();
+
+        assertThat(buffer.readUtf8()).isEqualTo("[\"map\",2,[\"str\",\"i32\",1,{\"1\":10}],[\"str\",\"i32\",1,{\"2\":20}]]");
+    }
+
+    @Test
+    public void structs() throws Exception {
+        Xtruct xtruct = new Xtruct.Builder()
+                .byte_thing((byte) 1)
+                .double_thing(2.0)
+                .i32_thing(3)
+                .i64_thing(4L)
+                .string_thing("five")
+                .build();
+
+        Xtruct.ADAPTER.write(protocol, xtruct);
+
+        assertThat(buffer.readUtf8()).isEqualTo("" +
+                "{" +
+                "\"1\":{\"str\":\"five\"}," +
+                "\"4\":{\"i8\":1}," +
+                "\"9\":{\"i32\":3}," +
+                "\"11\":{\"i64\":4}," +
+                "\"13\":{\"dbl\":2.0}" +
+                "}");
+    }
+
+    @Test
+    public void binary() throws Exception {
+        protocol.writeBinary(ByteString.encodeUtf8("foobar"));
```

Comment:
I think we should exercise the base-64 coder and decoder more thoroughly here.  Let's generate an array of bytes from 0..255 and encode that, so that we hit all of the encoding/decoding table entries.

I'm assuming that those are copied verbatim from Apache, and so are almost certainly correct, but we shouldn't assume such things in tests :)

Reply:
Yes, your assumption is correct. And yeah, I'll add tests for the encoding and decoding.